File: apps/shell/electron/coordinator/blackboard.ts
// Blackboard file management. Port of spike v3's resetBlackboard +
// readBlackboard + emptyBlackboard — with the v3 bug fixes applied
// (review_claude/gemini/merged NOT in initial schema; they're added by
// the coordinator after agent-4 runs, so agent-4 can't pre-fill them).

import { promises as fs } from 'node:fs';
import path from 'node:path';
import type { Blackboard, StagedInputs } from './types';

export const AGENTS = ['agent_1', 'agent_2', 'agent_3', 'agent_4'] as const;

const DEFAULT_PARAMS = {
  target_item_count: 6,
  difficulty_distribution: { easy: 0.34, medium: 0.5, hard: 0.16 },
  item_types: { mc_single: 0.5, fill: 0.3, ordering: 0.2 },
};

export function buildEmptyBlackboard(staged: StagedInputs): Blackboard {
  return {
    workflow: {
      current_step: 0,
      total_steps: 4,
      steps: [...AGENTS],
      status: 'pending',
    },
    user_input: {
      material: staged.material,
      competency_dimensions: staged.dimensions ?? [],
      domain_guidance: staged.domain_guidance,
      assessment_params: staged.assessment_params ?? DEFAULT_PARAMS,
    },
    data: {
      knowledge_units: null,
      mapping: null,
      items: null,
      review: null,
    },
    log: [],
    costs: { total_usd: 0, by_agent: {} },
  };
}

export async function writeBlackboard(filePath: string, board: Blackboard): Promise<void> {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, JSON.stringify(board, null, 2));
}

export async function readBlackboard(filePath: string): Promise<Blackboard | null> {
  try {
    const raw = await fs.readFile(filePath, 'utf-8');
    return JSON.parse(raw) as Blackboard;
  } catch {
    return null;
  }
}

export async function resetBlackboard(filePath: string, staged: StagedInputs): Promise<Blackboard> {
  const b = buildEmptyBlackboard(staged);
  await writeBlackboard(filePath, b);
  return b;
}

File: apps/quiz/src-tauri/src/blackboard.rs
// Blackboard file management — port of
// apps/shell/electron/coordinator/blackboard.ts.
// `default_params` / `build_empty_blackboard` / `write_blackboard` /
// `reset_blackboard` are only called from the workflow port (Phase 4).

#![allow(dead_code)]

use std::collections::HashMap;
use std::path::Path;
use tokio::fs;

use crate::types::{
    AssessmentParams, Blackboard, Costs, DataSlice, DifficultyDistribution, StagedInputs,
    UserInput, WorkflowState, WorkflowStatus,
};

pub const AGENTS: [&str; 4] = ["agent_1", "agent_2", "agent_3", "agent_4"];

pub fn default_params() -> AssessmentParams {
    AssessmentParams {
        target_item_count: 6,
        difficulty_distribution: DifficultyDistribution {
            easy: 0.34,
            medium: 0.5,
            hard: 0.16,
        },
        item_types: Some(HashMap::from([
            ("mc_single".to_string(), 0.5),
            ("fill".to_string(), 0.3),
            ("ordering".to_string(), 0.2),
        ])),
        item_type_counts: None,
    }
}

pub fn build_empty_blackboard(staged: &StagedInputs) -> Blackboard {
    Blackboard {
        workflow: WorkflowState {
            current_step: 0,
            total_steps: 4,
            steps: AGENTS.iter().map(|s| s.to_string()).collect(),
            status: WorkflowStatus::Pending,
        },
        user_input: UserInput {
            material: staged.material.clone(),
            competency_dimensions: staged.dimensions.clone().unwrap_or_default(),
            domain_guidance: staged.domain_guidance.clone(),
            assessment_params: staged
                .assessment_params
                .clone()
                .unwrap_or_else(default_params),
        },
        data: DataSlice::default(),
        log: vec![],
        costs: Costs::default(),
    }
}

pub async fn write_blackboard(path: &Path, board: &Blackboard) -> Result<(), String> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)
            .await
            .map_err(|e| format!("mkdir {}: {}", parent.display(), e))?;
    }
    let json = serde_json::to_string_pretty(board).map_err(|e| e.to_string())?;
    fs::write(path, json)
        .await
        .map_err(|e| format!("write {}: {}", path.display(), e))?;
    Ok(())
}

pub async fn read_blackboard(path: &Path) -> Option<Blackboard> {
    let raw = fs::read_to_string(path).await.ok()?;
    serde_json::from_str::<Blackboard>(&raw).ok()
}

pub async fn reset_blackboard(
    path: &Path,
    staged: &StagedInputs,
) -> Result<Blackboard, String> {
    let b = build_empty_blackboard(staged);
    write_blackboard(path, &b).await?;
    Ok(b)
}
